### PR TITLE
Implement a common set of algorithm semantics

### PR DESCRIPTION
**Pull request by stephenswat**
Currently, our algorithms are not really unified in any way; there is no common interface for using them. This commit implements a new interface which presents a unified set of semantics for algorithms, defined in terms of input and output parameters. Inspired by #31.

This is currently a proposal, open for comments. :slightly_smiling_face:

### CODE REVIEW

**Comment by paulgessinger** [comment]
Ok, so the combination of templates for the base class and virtual operators means that type erasure is confined to each 'type' of algorithm (i.e. seed finding). If that's the intention, then this looks good to me.

File: core/include/algorithms/component_connection.hpp [hunk 1]
```diff
@@ -76,8 +86,11 @@ namespace traccc {
         /// c++20 piping interface:
         /// @return a cluster collection
         ///
-        cluster_collection operator()(const device_cell_collection& cells,
-                                      const cell_module& module) const {
+        cluster_collection operator()(
+            const std::pair<const device_cell_collection &, const cell_module &> & i
+        ) const override {
+            const device_cell_collection& cells = i.first;
+            const cell_module& module = i.second;
```
[the remark this answers is not in the document]
Comment (stephenswat) [reply core/include/algorithms/component_connection.hpp:93]: I know, but I am severely allergic to the auto keyword and I prefer the more explicit version. If you want I change it over, though

**Comment by paulgessinger** [comment]
> Having to use composite objects (like `std::pair`) for input when the algorithm needs multiple objects is not necessarily the greatest. But I also don't have a better idea of how to do the abstraction. (At least without implementing a much bigger infrastructure. Again, like what LHCb did...)

I suppose CRTP with a 

```cpp
template <typename... Args>
output_t operator(Args&&... inputs);
```

that creates a tuple and forwards would work. But that wouldn't achieve perfect forwarding, and the output type would still have to be a tuple, so not really any clearer.

**Comment by stephenswat** [comment]
I must say I think a functional approach is ideal for what we are doing in traccc, and if you squint your eyes just a little you can actually start to see a lot of categorical structure which lends itself well to the functional style. If we consider our EDM to form a category **EDM** with objects representing abstract types and morphisms representing abstract algorithms, we can imagine our models of computing as functors from **EDM** to a category of concrete types **CPP**. For example, we can think of the entire CPU processing model as a functor *H* from **EDM** to **CPP** which transforms each abstract type in **EDM** into a concrete type for use on the host, and each morphism into a new concrete CPU algorithm. Similarly we can think of the whole CUDA processing model as a functor *D* from **EDM** to **CPP** which does virtually the same thing. Then we can even connect host and device processing together, because the act of copying memory from the host to the device and vice versa meets the requirement of a natural transformation *η*! As far as I can tell this meets all the categorical requirements for functors and natural transformations.

functor

The fact that we have a structure like this will allow us, I think, a _lot_ of powerful abstractions. The only question is how to encode that abstraction in the C++ programming language. But throwing away this structure would really be a waste, I think.

**Comment by cgleggett** [comment]
I'm not saying it's not a good pattern. Just that it doesn't mesh well with the existing Athena framework, and will  make things much more challenging if/when we integrate ACTS/traccc with Athena. Now of course ACTS != ATLAS, and it is good to have generic appeal, but we might be painting ourselves into a corner with this approach.

**Review by paulgessinger — APPROVED** [review]

**Comment by paulgessinger** [comment]
@stephenswat could you rebase? I merged #39.

Apart from this, if there are no additional discussion points, I don't see why we shouldn't go for these semantics for traccc. From what I can tell, we don't need to make the algorithm any more sophisticated.

**Comment by stephenswat** [comment]
I have finally rebased this merge request onto the re-formatted main branch, so I think we should be good to go. :smile:

**Review by paulgessinger — APPROVED** [review]

**Pull request merged**